RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu

```diff
@@ -0,0 +1,417 @@
+                                    ::ecal::reco::StorageScalarType* chi2sEE,
+                                    ::ecal::reco::StorageScalarType* energiesEB,
+                                    ::ecal::reco::StorageScalarType* energiesEE,
+                                    char* acState,
+                                    int nchannels,
+                                    int max_iterations,
+                                    uint32_t const offsetForHashes,
+                                    uint32_t const offsetForInputs) {
+      // FIXME: ecal has 10 samples and 10 pulses....
+      // but this needs to be properly treated and renamed everywhere
+      constexpr auto NSAMPLES = SampleMatrix::RowsAtCompileTime;
+      constexpr auto NPULSES = SampleMatrix::ColsAtCompileTime;
+      static_assert(NSAMPLES == NPULSES);
+
+      using DataType = SampleVector::Scalar;
+
+      extern __shared__ char shrmem[];
+      DataType* shrMatrixLForFnnlsStorage =
+          reinterpret_cast<DataType*>(shrmem) + calo::multifit::MapSymM<DataType, NPULSES>::total * threadIdx.x;
+      DataType* shrAtAStorage = reinterpret_cast<DataType*>(shrmem) +
+                                calo::multifit::MapSymM<DataType, NPULSES>::total * (threadIdx.x + blockDim.x);
+
+      // FIXME: remove eitehr idx or ch -> they are teh same thing
+      int idx = threadIdx.x + blockDim.x * blockIdx.x;
+
+// ref the right ptr
+#define ARRANGE(var) auto* var = idx >= offsetForInputs ? var##EE : var##EB
+      ARRANGE(amplitudes);
+      ARRANGE(chi2s);
+      ARRANGE(energies);
+#undef ARRANGE
+
+      auto const ch = idx;
+      if (idx < nchannels) {
+        if (static_cast<MinimizationState>(acState[idx]) == MinimizationState::Precomputed)
+          return;
+
+        // get the hash
+        int const inputCh = ch >= offsetForInputs ? ch - offsetForInputs : ch;
+        auto const* dids = ch >= offsetForInputs ? dids_ee : dids_eb;
+        auto const did = DetId{dids[inputCh]};
+        auto const isBarrel = did.subdetId() == EcalBarrel;
+        auto const hashedId = isBarrel ? ecal::reconstruction::hashedIndexEB(did.rawId())
+                                       : offsetForHashes + ecal::reconstruction::hashedIndexEE(did.rawId());
+
+        // inits
+        int iter = 0;
+        int npassive = 0;
+
+        calo::multifit::ColumnVector<NPULSES, int> pulseOffsets;
+#pragma unroll
+        for (int i = 0; i < NPULSES; ++i)
+          pulseOffsets(i) = i;
+
+        calo::multifit::ColumnVector<NPULSES, DataType> resultAmplitudes;
+#pragma unroll
+        for (int counter = 0; counter < NPULSES; counter++)
+          resultAmplitudes(counter) = 0;
+
+        // inits
+        //SampleDecompLLT covariance_decomposition;
+        //SampleMatrix inverse_cov;
+        SampleVector::Scalar chi2 = 0, chi2_now = 0;
+
+        // loop until ocnverge
+        while (true) {
+          if (iter >= max_iterations)
```

Comment:
I would encourage to wrap this rather long loop body in a function. Now with all the spacing and pragmas, it's quite hard to grasph.